File: L4/Tree/Tree.cpp
#include <iostream>
#include <stdexcept>
#include "Tree.h"

template<typename T>
void Tree<T>::addRec(T value, TreeNode<T>* parent) {
    if (value > parent->getValue()) {
        if (parent->getRightChild() != nullptr) {
            this->addRec(value, parent->getRightChild());
        } else {
            parent->setRightChild(value);
        }
    } else {
        if (parent->getLeftChild() != nullptr) {
            this->addRec(value, parent->getLeftChild());
        } else {
            parent->setLeftChild(value);
        }
    }
}

template<typename T>
void Tree<T>::showInOrderRec(TreeNode<T>* TreeNode) {
    if (TreeNode->getLeftChild() != nullptr) {
        showInOrderRec(TreeNode->getLeftChild());
    }
    std::cout << TreeNode->getValue() << " ";

    if (TreeNode->getRightChild() != nullptr) {
        showInOrderRec(TreeNode->getRightChild());
    }
}

template<typename T>
bool Tree<T>::findRec(T value, TreeNode <T>* TreeNode) {
    if (TreeNode == nullptr) {
        numberOfOperationsInFind++;
        return false;
    } else if (TreeNode->getValue() < value) {
        numberOfOperationsInFind++;
        return findRec(value, TreeNode->getRightChild());
    } else if (TreeNode->getValue() > value) {
        numberOfOperationsInFind++;
        return findRec(value, TreeNode->getLeftChild());
    } else {
        return true;
    }
}

template<typename T>
Tree<T>::Tree() {
    root = nullptr;
    treeSize = 0;
    numberOfOperationsInFind = 0;
};


template<typename T>
void Tree<T>::insert(T value) {
    treeSize++;
    if (root == nullptr) {
        root = new TreeNode<T>(value, nullptr);
    } else {
        this->addRec(value, root);
    }
}

template<typename T>
void Tree<T>::inorder() {
    if (root == nullptr) throw std::runtime_error("Tree is empty");
    showInOrderRec(root);
}

template<typename T>
bool Tree<T>::find(T value) {
    return this->findRec(value, root);
}

template<typename T>
int Tree<T>::min() {
    if (root == nullptr) throw std::runtime_error("Tree is empty");
    TreeNode<T>* minimalNode = findMinimum(root);
    return minimalNode->getValue();
}

template<typename T>
int Tree<T>::max() {
    if (root == nullptr) throw std::runtime_error("Tree is empty");
    TreeNode<T>* TreeNode = root;
    while (TreeNode->getRightChild() != nullptr) {
        TreeNode = TreeNode->getRightChild();
    }

    return TreeNode->getValue();
}

template<typename T>
int Tree<T>::size() {
    return treeSize;
}

template<typename T>
void Tree<T>::remove(T value) {
    treeSize--;
    deleteRec(value, root);
}

template<typename T>
void Tree<T>::deleteRec(T value, TreeNode <T>* treeNode) {
    if (treeNode == nullptr) {
        return;
    } else if (treeNode->getValue() < value) {
        deleteRec(value, treeNode->getRightChild());
    } else if (treeNode->getValue() > value) {
        deleteRec(value, treeNode->getLeftChild());
    } else {
        deleteOperation(treeNode);
    }
}

template<typename T>
void Tree<T>::deleteOperation(TreeNode <T>* treeNode) {
    if (treeNode == root) {
        if (treeNode->getRightChild() == nullptr && treeNode->getLeftChild() == nullptr) {
            root = nullptr;
            delete treeNode;
        } else if (treeNode->getRightChild() == nullptr) {
            root = treeNode->getLeftChild();
            delete treeNode;

        } else if (treeNode->getLeftChild() == nullptr) {
            root = treeNode->getRightChild();
            delete treeNode;

        } else {
            TreeNode<T>* smallestInRightSubTree = findMinimum(treeNode->getRightChild());
            changeParentsChildren(smallestInRightSubTree, nullptr);
            treeNode->setValue(smallestInRightSubTree->getValue());
            delete smallestInRightSubTree;
        }
        return;
    } else if (treeNode->getRightChild() == nullptr && treeNode->getLeftChild() == nullptr) {
        changeParentsChildren(treeNode, nullptr);
    } else if (treeNode->getRightChild() == nullptr) {
        changeParentsChildren(treeNode, treeNode->getLeftChild());
    } else if (treeNode->getLeftChild() == nullptr) {
        changeParentsChildren(treeNode, treeNode->getRightChild());
    } else {
        TreeNode<T>* smallestInRightSubTree = findMinimum(treeNode->getRightChild());
        changeParentsChildren(smallestInRightSubTree, nullptr);
        treeNode->setValue(smallestInRightSubTree->getValue());
        delete smallestInRightSubTree;
    }
    delete treeNode;
}

template<typename T>
void Tree<T>::changeParentsChildren(TreeNode <T>* treeNode, TreeNode <T>* newChildren) {
    if (treeNode->getParent()->getValue() <= treeNode->getValue()) {
        treeNode->getParent()->changeRightChild(newChildren);
    } else {
        treeNode->getParent()->changeLeftChild(newChildren);
    }
}

template<typename T>
TreeNode<T>* Tree<T>::findMinimum(TreeNode <T>* treeRoot) {
    TreeNode<T>* minimumNode = treeRoot;
    while (minimumNode->getLeftChild() != nullptr) {
        minimumNode = minimumNode->getLeftChild();
    }
    return minimumNode;
}

template<typename T>
int Tree<T>::getNumberOfOperationsInFind() {
    return numberOfOperationsInFind;
}

template<typename T>
void Tree<T>::resetNumberOfOperationsInFind() {
    numberOfOperationsInFind = 0;
}

template
class Tree<int>;

File: L5/Tree.py
from enum import Enum


class Tree:
    def __init__(self):
        self.root = None
        self.size = 0
        self.operations = 0

    def length(self):
        return self.size

    def __len__(self):
        return self.size

    def __iter__(self):
        return self.root.__iter__()

    def insert(self, value):
        if self.root is None:
            self.root = TreeNode(value, None, self)
        else:
            self.root.add(value)

    def show(self):
        if self.root is None:
            raise Exception("Tree is empty")
        else:
            self.root.show()

    def min(self):
        return self.root.min().value

    def max(self):
        return self.root.max().value

    def delete(self, value):
        if self.root is None:
            raise Exception("No such element")
        else:
            self.root.delete(value)

    def find(self, value):
        return 1 if self.root.find(value) else 0

    def os_select(self, x, i):
        if x is None:
            raise Exception("No such element")
        elif x.left is None:
            r = 1
        else:
            r = x.left.size + 1
        if i == r:
            return x
        elif i < r:
            return self.os_select(x.left, i)
        else:
            return self.os_select(x.right, i - r)

    def os_rank(self, x):
        if x.left is None:
            r = 1
        else:
            r = x.left.size + 1
        y = x
        while y != self.root:
            if y == y.parent.right:
                if y.parent.left is None:
                    r = r + 1
                else:
                    r = r + y.parent.left.size + 1
            y = y.parent
        return r

    def clear_operations_counter(self):
        self.operations = 0


class TreeNode:
    def __init__(self, value, parent, tree):
        self.value = value
        self.left = None
        self.right = None
        self.parent = parent
        self.tree = tree
        self.size = 1

    def add(self, value):
        self.size += 1
        if value > self.value:
            if self.right is None:
                self.right = TreeNode(value, self, self.tree)
            else:
                self.right.add(value)
        else:
            if self.left is None:
                self.left = TreeNode(value, self, self.tree)
            else:
                self.left.add(value)

    def show(self):
        if self.left is not None:
            self.left.show()
        print(self.value)
        if self.right is not None:
            self.right.show()

    def min(self):
        if self.left is None:
            return self
        else:
            return self.left.min()

    def max(self):
        if self.right is None:
            return self
        else:
            return self.right.max()

    def delete(self, value):
        if self.value == value:
            if self.left is None and self.right is None:
                if self == self.tree.root:
                    self.tree.root = None
                elif self.parent.value > self.value:
                    self.parent.left = None
                else:
                    self.parent.right = None
            elif self.left is None:
                if self == self.tree.root:
                    self.tree.root = self.right
                    self.right.parent = None
                elif self.parent.value > self.value:
                    self.parent.left = self.right
                    self.right.parent = self.parent
                else:
                    self.parent.right = self.right
                    self.right.parent = self.parent
            elif self.right is None:
                if self == self.tree.root:
                    self.tree.root = self.left
                    self.left.parent = None
                elif self.parent.value > self.value:
                    self.parent.left = self.left
                    self.left.parent = self.parent
                else:
                    self.parent.right = self.left
                    self.left.parent = self.parent
            else:
                self.size -= 1
                t = self.right
                while t.left is not None:
                    t.size -= 1
                    t = t.left
                self.value = t.value
                t.delete(t.value)
        else:
            if self.value > value:
                if self.left is None:
                    raise Exception("No such an element")
                else:
                    self.size -= 1
                    self.left.delete(value)
            else:
                if self.right is None:
                    raise Exception("No such an element")
                else:
                    self.size -= 1
                    self.right.delete(value)

    def find(self, value):
        if self.value == value:
            return True
        elif self.value > value:
            if self.left is None:
                return False
            else:
                return self.left.find(value)
        else:
            if self.right is None:
                return False
            else:
                return self.right.find(value)
